pilot/pkg/model/credentials/resource.go

```diff
@@ -0,0 +1,100 @@
+	KubernetesGatewaySecretType    = "kubernetes-gateway"
+	kubernetesGatewaySecretTypeURI = KubernetesGatewaySecretType + "://"
+)
+
+// SecretResource defines a reference to a secret
+type SecretResource struct {
+	// Type is the type of secret. One of KubernetesSecretType or KubernetesGatewaySecretType
+	Type string
+	// Name is the name of the secret
+	Name string
+	// Namespace is the namespace the secret resides in. For implicit namespace references (such as in KubernetesSecretType),
+	// this will be resolved to the appropriate namespace. As a result, this should never be empty.
+	Namespace string
+	// ResourceName is the original name of the resource
+	ResourceName string
+	// Cluster is the cluster the secret should be fetched from.
+	Cluster cluster.ID
+}
+
+func (sr SecretResource) Key() string {
+	return "sds://" + sr.Type + "/" + sr.Name + "/" + sr.Namespace + "/" + string(sr.Cluster)
+}
+
+func ToKubernetesGatewayResource(namespace, name string) string {
+	return fmt.Sprintf("%s://%s/%s", KubernetesGatewaySecretType, namespace, name)
+}
+
+// ToResourceName turns a `credentialName` into a resource name used for SDS
+func ToResourceName(name string) string {
+	// If they explicitly defined the type, keep it
+	if strings.HasPrefix(name, kubernetesSecretTypeURI) || strings.HasPrefix(name, kubernetesGatewaySecretTypeURI) {
+		return name
+	}
+	// Otherwise, to kubernetes://
+	return kubernetesSecretTypeURI + name
+}
+
+// ParseResourceName parses a raw resourceName string.
+func ParseResourceName(resourceName string, proxyNamespace string, proxyCluster cluster.ID, configCluster cluster.ID) (SecretResource, error) {
+	sep := "/"
+	if strings.HasPrefix(resourceName, kubernetesSecretTypeURI) {
+		// Valid formats:
+		// * kubernetes://secret-name
+		// * kubernetes://secret-namespace/secret-name
+		// If namespace is not set, we will fetch from the namespace of the proxy. The secret will be read from
+		// the cluster the proxy resides in. This mirrors the legacy behavior mounting a secret as a file
+		res := strings.TrimPrefix(resourceName, kubernetesSecretTypeURI)
+		split := strings.Split(res, sep)
+		namespace := proxyNamespace
+		name := split[0]
+		if len(split) > 1 {
+			namespace = split[0]
+			name = split[1]
+		}
+		return SecretResource{Type: KubernetesSecretType, Name: name, Namespace: namespace, ResourceName: resourceName, Cluster: proxyCluster}, nil
+	} else if strings.HasPrefix(resourceName, kubernetesGatewaySecretTypeURI) {
+		// Valid formats:
+		// * kubernetes-gateway://secret-namespace/secret-name
+		// Namespace is required. The secret is read from the config cluster; this is the primary difference from KubernetesSecretType.
+		res := strings.TrimPrefix(resourceName, kubernetesGatewaySecretTypeURI)
+		split := strings.Split(res, sep)
+		if len(split) <= 1 {
+			return SecretResource{}, fmt.Errorf("invalid resource name %q. Expected namespace and name", resourceName)
+		}
+		namespace := split[0]
```

Comment:
what about empty cases "kubernetes-gateway:///name" or "kuberntes-gateway://ns/"

Reply:
Added a test case and error for these cases. Current callers of ParseResourceName appropriate handle errors, so I think its safe to make this return an error for these cases.